docs/websites/varnish/use-varnish-with-nginx-to-serve-cached-wordpress-content-over-ssl-on-debian-8.md

```diff
@@ -0,0 +1,683 @@
+		~~~
+
+23. After saving and exiting the file you have to reload the systemd process:
+
+        sudo systemctl daemon-reload
+
+24. To make Varnish start at boot up, just run the following command:
+
+		sudo systemctl enable varnish.service
+
+We won't start Varnish at this point because we need to configure nginx first.
+
+## Install and Configure PHP
+
+Before configuring nginx we have to install PHP-FPM. FPM is short for FastCGI Process Manager, and it allows the web server to act as a proxy passing all requests with the `.php` file extension to the PHP interpreter.
+
+1. To install PHP-FPM just run:
+
+        sudo apt-get install php5-fpm php5-mysql
+
+2. Open the `/etc/php5/fpm/php.ini` file to configure PHP-FPM:
+
+		sudo nano /etc/php5/fpm/php.ini
+
+3. Find the directive `cgi.fix_pathinfo=` and set it to `0`, for safety reasons. If this parameter is set to `1`, the PHP interpreter will try to process the file who's path is as close to the requested path as possible, if it's set to `0`, the interpreter will only process the file with the exact path, which is a safer alternative.
+
+		cgi.fix_pathinfo=0
+
+4. Now open `www.conf`:
+
+		sudo nano /etc/php5/fpm/pool.d/www.conf
+
+5. Edit the `listen =` directive to specify the socket that will be used by nginx to pass requests to PHP-FPM:
+
+		listen = /var/run/php5-fpm.sock
+
+6.  Restart PHP-FPM:
+
+		sudo systemctl restart php5-fpm
+
+7.  Now open `/etc/nginx/fastcgi_params`:
+
+		sudo nano /etc/nginx/fastcgi_params
+
+8.  Find the line `fastcgi_param  HTTPS  $https if_not_empty;`. Below it add the following two lines, necessary for nginx to interact with the FastCGI service:
+
+	{: .file-excerpt }
+	/etc/nginx/fastcgi_params
+	:   ~~~ nginx
+
+		fastcgi_param  SCRIPT_FILENAME    $request_filename;
+		fastcgi_param  PATH_INFO          $fastcgi_path_info;
+		~~~
+
+	Save and exit the file.
+
+## Configure nginx
+
+1.  Open `/etc/nginx/nginx.conf`:
+
+		sudo nano /etc/nginx/nginx.conf
+
+2. Comment out the following two lines, because we'll include these SSL settings in the server block within the `/etc/nginx/sites-enabled/default` file:
+
+	{: .file-excerpt }
+	/etc/nginx/nginx.conf
+	:   ~~~ nginx
+
+		# ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # Dropping SSLv3, ref: POODLE
+		# ssl_prefer_server_ciphers on;
+		~~~
+
+3. By default nginx makes use of an access log format called 'combined.' This type of logging is used by many web servers because it records all the information that is generally required by log analysis tools. The access log format can be changed from default to a custom one by using the `log_format` directive, but since we don't need custom logging we'll let nginx use the default 'combined' option.
+
+    Since the access logs and error logs will be defined for each individual website in the server block, you will also need to comment out the two lines below:
+
+	{: .file-excerpt }
+	/etc/nginx/nginx.conf
+	:   ~~~ nginx
+	
+		# access_log /var/log/nginx/access.log;
```

Comment:
Very interesting point. But here is the reason for my settings:
If we leave these two lines uncommented, Nginx will store in the `/var/log/nginx/access.log` file all the requests for all the https websites that we configured. It will also store all the errors for all the https websites in `/var/log/nginx/error.log`. If we have let's say three https sites configured in `/etc/nginx/sites-available/default`, Nginx will mix up all the requests to these three https websites in the access log. It will simply add the requests one below the other in the same log file: `/var/log/nginx/access.log`. It will not keep requests for different sites in different log files. We don't want this because this will make log analysis by tools such as Awstats or Webalizer impossible and even for occasional visuall checking of web traffic to each site. So, we want Nginx to write a different access log for each individual site. And the same for error logs. To do this we have to disable these general log settings and enable them in each server block for each https site individually.

For http websites things are indeed different. As I pointed out in the last paragraph of the guide, for http websites the first to receive the requests will not be Nginx but Varnish. So, there is no reason to enable Nginx access logging for http websites, either at the server block level or at the general level in `/etc/nginx/nginx.conf`, unless we want inaccurate logs. Nginx will log only the requests that will come from Varnish for those pages that were not found in cache and since most pages are found in cache, Nginx will only log a fraction of all the requests. If we want to have all the requests logged for http websites, we have to tell Varnish to log the requests for these sites. Varnish has a more complicated way of logging. It does it by starting a separate instance of a daemon called 'varnishncsa' for each individual website. The configuration of this kind of logging was too long to include in my guide. There are a few online resources that users can use to configure Varnish access logging. Since I coudn't include all this configuration, I just pointed out in the last paragraph:

> A step further can be to enable Varnish logging for the plain HTTP websites, since now Varnish is the first to receive the client requests, while nginx only receives requests for those pages that are not found in cache.

I now realize that it's good to include an error log directive in the Nginx configuration for the http site, since Varnish logs only access requests and it's recommended to also catch errors for individual http sites. So, if you have the time, please add this line before the last curly bracket in the second server block for the http website:

`error_log  /var/log/nginx/example-over-http.com.error.log notice;`